velero-plugins/imagestream/restore.go

```diff
@@ -18,22 +24,87 @@ func (p *RestorePlugin) AppliesTo() (velero.ResourceSelector, error) {
 	}, nil
 }
 
+// Execute copies local registry images from migration registry into target cluster local registry
 func (p *RestorePlugin) Execute(input *velero.RestoreItemActionExecuteInput) (*velero.RestoreItemActionExecuteOutput, error) {
 	p.Log.Info("Hello from ImageStream RestorePlugin!")
 
-	metadata, err := meta.Accessor(input.Item)
-	if err != nil {
-		return nil, err
-	}
-
-	annotations := metadata.GetAnnotations()
+	imageStream := imagev1API.ImageStream{}
+	itemMarshal, _ := json.Marshal(input.Item)
+	json.Unmarshal(itemMarshal, &imageStream)
+	p.Log.Info(fmt.Sprintf("image: %#v", imageStream))
+	annotations := imageStream.Annotations
 	if annotations == nil {
 		annotations = make(map[string]string)
 	}
 
-	annotations["openshift.io/imagestream-restore-plugin"] = "1"
+	imageStreamUnmodified := imagev1API.ImageStream{}
+	itemMarshal, _ = json.Marshal(input.ItemFromBackup)
+	json.Unmarshal(itemMarshal, &imageStreamUnmodified)
 
-	metadata.SetAnnotations(annotations)
+	internalRegistry := annotations[common.RestoreRegistryHostname]
+	if len(internalRegistry) == 0 {
+		return nil, errors.New("restore cluster registry not found for annotation \"openshift.io/restore-registry-hostname\"")
+	}
+	backupInternalRegistry := annotations[common.BackupRegistryHostname]
+	if len(internalRegistry) == 0 {
+		return nil, errors.New("backup cluster registry not found for annotation \"openshift.io/backup-registry-hostname\"")
+	}
+	migrationRegistry := input.Restore.Annotations[common.MigrationRegistry]
+	if len(migrationRegistry) == 0 {
+		return nil, errors.New("migration registry not found for annotation \"openshift.io/migration\"")
+	}
+	p.Log.Info(fmt.Sprintf("backup internal registry: %#v", backupInternalRegistry))
+	p.Log.Info(fmt.Sprintf("restore internal registry: %#v", internalRegistry))
 
-	return velero.NewRestoreItemActionExecuteOutput(input.Item), nil
+	for _, tag := range imageStreamUnmodified.Status.Tags {
+		p.Log.Info(fmt.Sprintf("tag: %#v", tag.Tag))
+		specTag := findSpecTag(imageStreamUnmodified.Spec.Tags, tag.Tag)
+		copyToTag := true
+		if specTag != nil && specTag.From != nil {
+			p.Log.Info(fmt.Sprintf("image tagged: %s, %s", specTag.From.Kind, specTag.From.Name))
+			// we have a tag.
+			copyToTag = false
+		}
+		// Iterate over items in reverse order so most recently tagged is copied last
+		for i := len(tag.Items) - 1; i >= 0; i-- {
+			dockerImageReference := tag.Items[i].DockerImageReference
+			localImage := strings.HasPrefix(dockerImageReference, backupInternalRegistry)
```

Comment:
Could optionally do:
```
if strings.HasPrefix(dockerImageReference, backupInternalRegistry) {
```

Reply:
I think in an earlier iteration of my backup plugin I used the variable more than once. This assignment isn't needed anymore.